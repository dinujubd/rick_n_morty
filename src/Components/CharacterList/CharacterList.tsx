import React, { useCallback, useState } from 'react';
import { connect } from 'react-redux';
import { Button, Grid, ListItem, ListItemAvatar, ListItemText, Avatar, List } from '@material-ui/core';
import { setCurretCharacterAction, loadMoreAction } from '../../Store/actions';
import { CharacterChunk } from '../../Store/characters';
import { ApplicationDispatch, ApplicationState } from '../../Store/store';

interface StateProps {
    characters: CharacterChunk[],
    nextUrl?: string,
}

interface DispatchProps {
    selectCharacter: (id: number) => void,
    loadMore: (url: string) => void
}

export const CharacterList: React.FC<StateProps & DispatchProps> = (props) => {

    const [selectedIndex, selectItem] = useState(1);

    const onClickHandler = useCallback((id: number) => () => {
        selectItem(id);
        props.selectCharacter(id);
    }, [selectItem, props])

    const onLoadMoreHandler = useCallback(() => {
        if (props.nextUrl)
            props.loadMore(props.nextUrl);
    }, [props])

    return (
        <List>
            {props.characters && props.characters.length > 0 ? props.characters.map(x => <ListItem key={x.id + x.name} button selected={selectedIndex === x.id} onClick={onClickHandler(x.id)}>
                <ListItemAvatar>
                    <Avatar src={x.image} />
                </ListItemAvatar>
                <ListItemText primary={x.name} />
            </ListItem>) : <ListItem> <Grid container justify="center">No characters found</Grid></ListItem>}
            {props.nextUrl && <ListItem>
                <Grid container justify="center">
                    <Button color="secondary" onClick={onLoadMoreHandler}>
                        Load More
                    </Button>
                </Grid>
            </ListItem>}
        </List>
    )
}

CharacterList.displayName = 'CharacterList';

const mapStateToProps = (state: ApplicationState): StateProps => ({
    characters: state.characters.characters,
    nextUrl: state.characters.nextUrl,
})

const mapDispatchToProps = (dispatch: ApplicationDispatch): DispatchProps => ({
    selectCharacter: (id: number) => dispatch(setCurretCharacterAction(id)),
    loadMore: (url: string) => dispatch(loadMoreAction(url))
})

export default connect(mapStateToProps, mapDispatchToProps)(CharacterList);