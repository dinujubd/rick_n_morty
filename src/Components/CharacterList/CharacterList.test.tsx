import * as React from 'react';
import renderer from 'react-test-renderer';
import { CharacterList } from './CharacterList';

describe('<CharacterList />', () => {
    it('should renders correctly', () => {
        const tree = renderer
            .create(<CharacterList characters={[]} selectCharacter={jest.fn()}  loadMore={jest.fn()}   />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});