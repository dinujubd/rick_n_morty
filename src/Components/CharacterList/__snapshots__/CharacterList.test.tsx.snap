// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CharacterList /> should renders correctly 1`] = `
<ul
  className="MuiList-root MuiList-padding"
>
  0
</ul>
`;
